multiset/__init__.py

```diff
@@ -13,8 +13,12 @@
 
 __all__ = ['BaseMultiset', 'Multiset', 'FrozenMultiset']
 
+_TElement = TypeVar('_TElement', bound=Hashable)
+_OtherType = Union[IterableType[_TElement], MappingType[_TElement, int]]
+_Self = TypeVar('_Self', bound='BaseMultiset')
+
 
-class BaseMultiset(object):
+class BaseMultiset(MappingType[_TElement, int], Generic[_TElement]):
```

Comment:
I have some concern with this solution, because as far as I remember, subclassing the typing types had some serious performance impacts every time you instantiate a new multiset. I don't know if that is still the case with more recent versions of the typings or not. But that was part of the reason for a separate type file.

Reply:
I'll run some benchmarks, didn't know that was the case!

Reply:
Reference: https://github.com/python/cpython/issues/83349

Reply:
Ok so I've ran some benchmarks using the pytest benchmark plugin, with the following code:
```python
from . import Multiset
from .multiset_wthout_generic import Multiset as MultisetWithoutGeneric


def subclassing_generic():
    return Multiset()


def not_subclassing_generic():
    return MultisetWithoutGeneric()


def test_subclassing_generic(benchmark):
    ms = benchmark(subclassing_generic)
    assert ms is not None


def test_not_subclassing_generic(benchmark):
    ms = benchmark(not_subclassing_generic)
    assert ms is not None

```
The `multiset_wthout_generic.py` contains the exact same code as my version of the multiset package except for the classes declarations which do not inherit from any typing class.

These are the results:
```
======================================================================================================= test session starts ========================================================================================================
platform win32 -- Python 3.12.1, pytest-8.0.0, pluggy-1.4.0
benchmark: 4.0.0 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: D:\lucad\Desktop\multiset
plugins: anyio-4.4.0, benchmark-4.0.0, django-4.8.0, time-machine-2.15.0
collected 346 items

multiset\test_benchmark.py ..                                                                                                                                                                                                 [  0%]
tests\test_multiset.py ...................................................................................................................................................................................................... [ 57%]
..................................................................................................................................................                                                                            [100%]


---------------------------------------------------------------------------------------------- benchmark: 2 tests ----------------------------------------------------------------------------------------------
Name (time in ns)                     Min                    Max                Mean              StdDev              Median                 IQR            Outliers  OPS (Mops/s)            Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_not_subclassing_generic     499.8874 (1.0)      40,399.9584 (1.46)     694.4358 (1.0)      489.6492 (1.0)      700.0053 (1.0)      100.0008 (1.0)     1555;3894        1.4400 (1.0)      158729           1
test_subclassing_generic         899.8904 (1.80)     27,599.9773 (1.0)      985.5509 (1.42)     640.1056 (1.31)     900.0069 (1.29)     100.0008 (1.0)      528;1722        1.0147 (0.70)      84033           1
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean
======================================================================================================= 346 passed in 2.31s ========================================================================================================
```


The time is in nanoseconds. My version is 1.42 times slower on average. I don't think this undermines the usability of this package, let me know what's your opinion on this.

Reply:
Thank you for doing the benchmarking! Looks okay to me, thank you for the PR.